File: algorithms/dynamic-programming/coin-change.py
class Solution:
  def coinChange(self, coins: List[int], amount: int) -> int:
    if amount == 0:
      return 0

    mem = [float('inf')] * (amount + 1)
    mem[0] = 0

    for coin in coins:
      for i in range(coin, amount + 1):
        mem[i] = min(mem[i], mem[i - coin] + 1)

    if mem[-1] == float('inf'):
      return -1
    else:
      return mem[-1]

File: top-questions/Google-Top-Questions/dynamic-programming/coin-change.js
var coinChange = function(coins, amount) {
  var mem = Array(coins.length).fill(0);

  for (var i = 0; i < coins.length; i++) {
    for (var j = coins[i]; j <= amount; j++) {
      mem[j] = min(mem[j], mem[j-coins[i]])
    }
  }

  return mem[amount]
};
